File: client/lib/src/features/chapters/domain/chapter.dart
import 'package:client/src/features/books/domain/book.dart';
import 'package:freezed_annotation/freezed_annotation.dart';

import '../../../common/ignore.dart';
import '../../comments/domain/comment.dart';

part 'chapter.freezed.dart';
part 'chapter.g.dart';

@freezed
class Chapter with _$Chapter {
  factory Chapter({
    @JsonKey(name: "_id") @Default("") String id,
    @Default("") String name,
    @Default("") String content,
    @Default(ReadingsState.unpublished) ReadingsState state,
    // @JsonKey(name: 'book') required String bookId,
    required Book book,
    @JsonKey(toJson: ignore, includeIfNull: false) List<Comment>? comments,

    DateTime? createdAt,
    DateTime? updatedAt
  }) = _Chapter;

  factory Chapter.fromJson(Map<String, dynamic> json) => _$ChapterFromJson(json);
}

chapterListFromJson(List<dynamic> list) {
  return list.map((c) => Chapter.fromJson(c as Map<String, dynamic>))
    .toList();
}

extension ChapterStateExtension on Chapter {
  bool get isPublished => state == ReadingsState.published;
  bool get isUnpublished => state == ReadingsState.unpublished;
}

extension ChapterLifecycleExtension on Chapter { 
  bool exists() => id.isNotEmpty;
}

File: admin/src/features/chapters/domain/chapter.ts
import { Book } from "features/books/domain/book"
import { BaseCommentSubject } from "features/comments/domain/comment-subject"

export class Chapter implements BaseCommentSubject {
  _id: string

  name: string
  content: string
  book: Book
  state: string
  
  createdAt?: Date
  updatedAt?: Date

  constructor(data: any) {
    this._id = data._id

    this.name = data.name
    this.content = data.content
    this.state = data.state
    this.book = new Book(data.book)

    if (data.createdAt) this.createdAt = new Date(data.createdAt)
    if (data.updatedAt) this.updatedAt = new Date(data.updatedAt)
  }
}
